### PR TITLE
.dockerignore ignore support. Fixes #9171

### LINKED ISSUE #9171: .dockerignore should a support including ignored content
If the only thing you want in your dockerfile is the Gemfile of a project it is tedious to use .dockerignore to exclude everything manually.

With .gitignore there is support for including things allowing you to do:

```
/*
!/Gemfile
!/Gemfile.lock
```
Having similar for dockerignore would be very nice.

more background: http://stackoverflow.com/questions/5533050/gitignore-exclude-folder-but-include-specific-subfolder

### PR DESCRIPTION
**Pull request by omeid**
Allow ignoring files from .dockerignore exclude lists like .gitingore by prefixing `!` to a rule.

Closes #9171

### CODE REVIEW

**Comment by jfrazelle** [comment]
Can you please sign your commits following these rules:

https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work

The easiest way to do this is to amend the last commit:

`git commit --amend -s --no-edit && git push -f`

**Comment by SvenDowideit** [comment]
This needs documentation - run `git grep dockerignore` to find where.

**Comment by tiborvass** [comment]
Ping @omeid

**Comment by omeid** [comment]
Okay, I will fix the signing in a few hours.

**Comment by duglin** [comment]
@omeid can you also add some tests

**Comment by icecrime** [comment]
Hello @omeid, thanks for the contribution! Could you please add unit tests for this new feature?

**Comment by omeid** [comment]
Sure, but where do I put the tests? the sub-packages doesn't seem to have independent testing and this is part of the "Builder", so is that where I should add extra cases for exclusion?

File: pkg/fileutils/fileutils.go [hunk 1]
```diff
@@ -6,21 +6,43 @@ import (
 )
 
 // Matches returns true if relFilePath matches any of the patterns
+// and isn't excluded by any of the patterns.
```
Comment (omeid) [inline pkg/fileutils/fileutils.go]: Note to self: This should be "// and isn't excluded by any _subsequent_ pattern."

**Comment by duglin** [comment]
add tests to integration-cli/docker_cli_build_test.go

**Comment by icecrime** [comment]
Sorry, I should have give guidance about that. Agree with @duglin, although ut's also perfectly fine to add unit tests in a new `pkg/fileutils/fileutils_test.go` file.

**Comment by buddhamagnet** [comment]
Just wondered whether it makes sense to make these changes in ReadDockerIgnore in utils.go?

**Comment by buddhamagnet** [comment]
@omeid could you give me push access to your docker repo would love to help on this.

**Comment by omeid** [comment]
@buddhamagnet Done. 

I tried to write tests but the [test file](integration-cli/docker_cli_build_test.go) is a massive over 4.5k lines and it slows down my current text editor setup to point of unusable. :/

**Comment by buddhamagnet** [comment]
@omeid added PR to your fork for comment clarification:

https://github.com/docker/docker/pull/11711

**Comment by buddhamagnet** [comment]
@omeid we could put the changes into ReadDockerIgnore in utils.go as that seems more localised? The beauty of that is we could add a unit test in utils_test.go?

**Comment by omeid** [comment]
The problem of putting this code there is that ReadDockerIgnore is basically a parser and, rightfully so, doesn't return files but list of the patterns, trying to evaluate the patterns there is out of the scope of the function.

The code is in the right place, I believe, and it will be consumed by `*Builder.readDockerfile`.

I am almost done with the tests just need a rebase of everything and there has been a lot of changes since this pull request.

**Comment by buddhamagnet** [comment]
@omeid sounds good I have only just started looking at contributing to this project.

**Comment by buddhamagnet** [comment]
@omeid we could add a fileutils_test.go test file in the same directory?

**Comment by duglin** [comment]
don't forget to add docs for this too.

**Comment by buddhamagnet** [comment]
@omeid I have a unit test completed, we can build on that. Happy for me to commit?

@duglin yep docs will be included.

**Comment by buddhamagnet** [comment]
@omeid I have three unit tests stacked up against fileutils.Matches, let me know how your rebase is going and I'll add them when you're done?

**Comment by buddhamagnet** [comment]
@omeid if you want I am happy to spawn a new branch direct from docker master and pipe these changes back in in a new PR with the tests if the rebase is getting hairy?

**Comment by buddhamagnet** [comment]
@omeid and have started adding documentation in:
```txt
docs/sources/articles/dockerfile_best-practices.md
docs/sources/reference/builder.md
docs/sources/reference/commandline/cli.md
```

**Comment by buddhamagnet** [comment]
@icecrime @tiborvass @duglin I'm having trouble getting a response from @omeid. I have unit tests against fileutils.Matches and documentation. Should I issue a new PR or wait? Don't want to step on any toes here!

**Comment by omeid** [comment]
It was my first request and a lot of things got a bit messy. I am closing this down, follow up on #11823.

@buddhamagnet  20 Hours isn't what I consider having trouble to get a response. You're making a lot of noise. Please slow down.

**Comment by buddhamagnet** [comment]
@omeid creating noise? I was trying to help, which is why I (believe politely) asked that question.

**Pull request closed without merging**